### ISSUE TITLE
+13% in startup time, augmentation time, and after-start memory footprint

### ISSUE THREAD

**Issue by tkyjovsk** [keycloak/keycloak#20448]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

core

### Describe the bug

CI job running with Keycloak Nightly images discovered a 12-14% increase in startup time, augmentation time, and memory footprint after start between 2023-04-28 and 2023-04-29 (both at 2:25 AM UTC). The regression has been present in all subsequent runs since then.
![Screenshot from 2023-05-18 17-11-37](https://github.com/keycloak/keycloak/assets/1867605/94310a65-5c14-4384-8f16-bf7526d664e9)

### Version

nightly

### Expected behavior

Server startup time and augmentation time (as reported in the server log) and memory footprint after start at the level of pre-Quarkus 3 merge baseline.

### Actual behavior

+12 .. 14% regression compared to baseline.

### How to Reproduce?

Measure startup time, augmentation time as reported in Keycloak server's pod log; memory usage as reported in `/sys/fs/cgroup/memory/memory.usage_in_bytes` or `/proc/1/smaps` inside the Keycloak pod. Utility scripts for this will be added to the keycloak-benchmark project shortly.

**Comment by hmlnarik** [comment]
Maybe related to #19919?

**Comment by mposolda** [comment]
This is 99.9% related to quarkus3 upgrade (as also mentioned in the description), so moving out of `core` team dashboard for now.

**Comment by mabartos** [comment]
@tkyjovsk @hmlnarik @mposolda Yes, definitely related to the Quarkus 3 upgrade. I'll analyze it and discuss it with our sub-team.

**Comment by geoand** [comment]
@franz1981 @Sanne you guys ran a bunch of tests around startup performance, correct?
As Keycloak does use Hibernate maybe this is related?

**Comment by franz1981** [comment]
@mabartos As suggested by @geoand - we do have an increase in both due to ANTL components class loading (and other parts re HB 6) - do you use Hibernate?

**Comment by Sanne** [comment]
ANTLR 4 based grammars are indeed a bit heaver than the ones generated by ANTLR v2 (we upgraded to v4 in Hibernate ORM 6, using v2 in ORM 5), and the difference was noticeable in some of our tests but in absolute terms it was "just" about ~50ms in my machine.

From reading the graph it seems the startup time regression here is in the order of 10 seconds? If I'm reading that right then I don't think this is caused by the ANTLR related regression. Could you report the differences in absolute numbers please?

Also, worth checking if the bootstrap times aren't deteriorating as a direct consequence of it consuming a bit more memory; could be worth plotting used heap sizes and GC activity and compare between the two versions.

Incidentally, I didn't expect Keycloak to need so much time to boot. Could you elaborate on what the process entails? Does it need to wait for external services to timeout/heartbeat, perform some extensive IO operations or is this time purely dictated by the runtime's ability to quickly load, initialize and compile classes?

**Comment by franz1981** [comment]
@Sanne I was missing the "seconds" on the axis - yep, we need a confirmation that it is the right unit of measure and what the test does

**Comment by tkyjovsk** [comment]
@Sanne Hello. Here you go:

| Metric                                            | Test Run 89 (2023-04-28) | Test Run 90 (2023-04-29) | DIFF     | DIFF in % |
|---------------------------------------------------|---------------------------|---------------------------|----------|-----------|
| Keycloak server augmentation time (s)             |                   41.7213 |                   47.2133 |    5.492 |    13.16% |
| **Keycloak server startup time (s)**                  |                   19.1486 |                   21.3714 |   2.2228 |    11.61% |
| Containers readying-time (s)                      |                      76.4 |                      84.2 |      7.8 |    10.21% |
| RSS at 120s after start (B)                           |                 316547482 |                 358298419 | 41750938 |    13.19% |
| CGroups Memory at 120s after start (B) |                 494532198 |                 560899277 | 66367078 |    13.42% |

Comparison with latest nightly build:

| Metric                                 | Test Run 89 (2023-04-28) | Test Run 120 (2023-05-29) | DIFF     | DIFF in % |
|----------------------------------------|---------------------------|----------------------------|----------|-----------|
| Keycloak server augmentation time (s)  |                   41.7213 |                    46.2614 |   4.5401 |    10.88% |
| **Keycloak server startup time (s)**       |                   19.1486 |                    21.5498 |   2.4012 |    12.54% |
| Containers readying-time (s)           |                      76.4 |                       83.1 |      6.7 |     8.77% |
| RSS at 120s after start (B)            |                 316547482 |                  353668710 | 37121229 |    11.73% |
| CGroups memory at 120s after start (B) |                 494532198 |                  556813517 | 62281318 |    12.59% |

**Comment by Sanne** [comment]
Very interesting, thanks!

So - I'm pretty sure that's not caused by the ANTLR issue as that's a micro issue in comparison - it's one of the few regressions we were aware of, but it's much, much smaller so couldn't possibly justify this.

I'm afraid we can't offer more blind guesses, we'll need to profile it. Are you familiar with producing flamegraphs or JFRs?
Either @franz1981 or me can help, in case, to setup for producing them. I'd certainly be interested to look at them - once we have them it should be straight forward to know what's happening.

**Comment by mabartos** [comment]
Thanks, guys, for your suggestions! I'll take a look at it this week.

**Comment by mabartos** [comment]
I've tried to measure the start-up and time needed for execution particular build steps for runtime with using the `-Dquarkus.debug.print-startup-times=true`. I was able to reproduce the performance decrease with the Quarkus 3 upgrade. I've only gathered the results, and I'll analyze the critical spots later.

I've tested a comparison between Keycloak 21.0.2 and the latest upstream main branch with changes added during the Quarkus 3 upgrade initiative.

<table>
    <tr>
       <td align="center">21.0.2 with JPA</td>
       <td align="center">latest with JPA</td>
    </tr>
    <tr>
      <td>
        <img src="https://github.com/keycloak/keycloak/assets/38039883/f892232a-c281-433d-b9a7-845bb04b3158" alt="First Image">
      </td>
      <td>
        <img src="https://github.com/keycloak/keycloak/assets/38039883/fcdcceb5-4dc3-47ec-b033-05d01696d53d" alt="Second Image">
      </td>
    </tr>
  </table>
  
As you can see the most significant performance decrease is in the:
**LifecycleEventsBuildStep.startupEvent
HibernateOrmProcessor.startPersistenceUnits**

It seems that the first one is related to the callback executed during observing the `StartupEvent`, which is basically this one in our codebase: https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/services/resources/KeycloakApplication.java#L124

I haven't analyzed it yet, but I'll do it. 
The other one is related to the starting persistence units. 

You can check the results for a start-up without using the JPA; used CHM (concurrent hash map).

<table>
    <tr>
       <td align="center">21.0.2 without JPA</td>
       <td align="center">latest without JPA</td>
    </tr>
    <tr>
      <td>
        <img src="https://github.com/keycloak/keycloak/assets/38039883/fdde803f-00bb-43da-b221-8c60069ea8e0" alt="First Image">
      </td>
      <td>
        <img src="https://github.com/keycloak/keycloak/assets/38039883/aaf6ecbe-9dd5-46e6-bed8-d418a100d4e2" alt="Second Image">
      </td>
    </tr>
  </table>

I'll know more details soon.

cc @tkyjovsk

**Comment by Sanne** [comment]
Thanks - interesting. So what I see is that clearly disabling JPA helps significantly, but at the same time the slowdown doesn't seem _specific_ to booting Hibernate: almost all different phases take much longer; it's actually surprising how much the slowdown is spread out almost homogeneously.

I'd suggest to keep an eye on GC activity; it's possible that a slightly higher memory consumption in the new components (anywhere) is getting you in a situation for which the amount of memory is a bit too restrained for optimal execution efficiency. Clearly disabling JPA would save you some memory, hence having a beneficial impact all around.

A similar theory could apply to other constrained resources, e.g. CPU; but I'd check GC logs first just because it's very easy to check and also quite possible.

**Comment by mabartos** [comment]
@franz1981 I'll analyze the results more and summarize my findings until tomorrow or the beginning of the following week. Thanks for that!

**Comment by franz1981** [comment]
Adding @barreiro to the party, he might be interested into the Agroal bits.
I will take a deeper look at the comment tomorrow or next week, but related Hibernate we have indeed detected some G1 consuming oprations, in particular the card marking scanning, which means both a huge amount of retained garbage or effectively some form of lean (including GC nepotism).
What could be a nice data to collect is the oldObject JFR events that can help found which classes are the mostly likely candidate for some leaks.
Given the continuos nature of parallel heap generations (which are not scattered as the G1 regions) it could be that scanning (that should still happen on it) won't have the same impact. To detect this one instead would be great to collect -e cache-misses events on GC C++ code using a proper native profiler like async-profiker, but not sure that a viable option unless we could run the same test outside a container.

**Comment by franz1981** [comment]
@mkouba for the Arc side instead

**Comment by mkouba** [comment]
> @mkouba for the Arc side instead

Well, the `io.quarkus.arc.runtime.ArcRecorder.initBeanContainer` itself does not perform any logic but instead invokes all the registered `io.quarkus.arc.runtime.BeanContainerListener`s.

In this particular case, it seems to be a perf regression in the  `io.quarkus.hibernate.orm.runtime.PersistenceUnitsHolder.initializeJpa()` method.

If you enable the `DEBUG` level for the `io.quarkus.arc.runtime.ArcRecorder` logger you should see a `Bean container listener %s finished in %s ms` message for each listener executed.

**Comment by mabartos** [comment]
@ahus1 Thanks for the info. As you're suggesting, I already tried to reduce the number of allocation failures by increasing the initial heap size, but it didn't help much. However, I can try to experiment with it a little bit more.

**Comment by franz1981** [comment]
@mabartos 

by converting the stack samples in both the JFR files for the comparison one thing immediately visible is an increase activity of defineClass1 while moving to Quarkus 3 (as mentioned in the previous comment):
![image](https://github.com/keycloak/keycloak/assets/13125299/43ac6738-559a-42f2-bb0b-0146763829c5)

on the left there's the quarkus 2 version, while on the right, the quarkus 3 one.
Given that samples are collected by JFR is not as straight to compare them but each of the violet top methods ie `defineClass1` weight 1 samples and quarkus 2 got 5 while quarkus 3 got 12, that's more then double activity and represent a large portion of the samples collected around JPA/Hibernate in the already mentioned previous comments.

In short: collecting class loading per type JFR events could help (given that they should bring both stack trace and duration of the event) or async-profiler, in 2 ways:  
- using the shiny new `--classify` mode which allow to get which portion of cpu time is spent in different type of JVM activities eg compilations,virtual calls, define/load classes, GC, C2 compiled code, C1 compiled code, Interpreted
- or via instrumentation per `-e Java_java_lang_ClassLoader_defineClass1` which can report how many times and which stack trace happen that goes into the native method used by the JDK to define the classes (here the sample count is the number of times the method is called, it's no longer "sampled" at intervals)

**Comment by geoand** [comment]
>  or async-profiler, in 2 ways:
> 
> * using the shiny new `--classify` mode which allow to get which portion of cpu time is spent in different type of JVM activities eg compilations,virtual calls, define/load classes, GC, C2 compiled code, C1 compiled code, Interpreted
> * or via instrumentation per `-e Java_java_lang_ClassLoader_defineClass1` which can report how many times and which stack trace happen that goes into the native method used by the JDK to define the classes (here the sample count is the number of times the method is called, it's no longer "sampled" at intervals). See the doc at https://github.com/async-profiler/async-profiler#java-method-profiling

Very cool, I did not know about these!

**Comment by vmuzikar** [comment]
@mabartos @tkyjovsk I wonder what's the current status of this. IIRC, the perf regression was not on our side in the end. Can we close this then?

**Comment by franz1981** [comment]
@mabartos any news on it?

**Comment by mabartos** [comment]
@franz1981 Sorry, I haven't had a chance to investigate it more extensively. Have you found any news on your side? 

AFAIK, we don't have any continuous performance measures aggregator ATM, so we are not able to compare the situation now and before, for the distribution. It means the validity of this issue is decreasing every day :)) Perhaps @tkyjovsk can add more details for it.

For that reason, I'll close the issue marked as 'bug' and open a new issue for analyzing potential performance pitfalls in Keycloak as an enhancement issue.
 
If anyone is against it, we can discuss it more.

**Comment by franz1981** [comment]
Yep @mabartos we have improved recently startup time (with Roberto Cortez) quite a lot so you'll probably get it from >3.4.1 version and we will see if it's a relevant improvement for you as well...

**Comment by Sanne** [comment]
It's cool, but please let us know when you have more metrics!
We all stand to learn a lot from such work and I'm keen to improve both Keycloak and Quarkus, if we can..
